examples/extra_element_integer/clustering_algorithm.h

```diff
@@ -0,0 +1,149 @@
+        equation_systems.reinit();
+    }
+}
+
+double GetElementDataFromMesh(LinearImplicitSystem &system,
+                              const libMesh::Elem &elem,
+                              const unsigned int variable_num) {
+
+    DofMap &dof_map = system.get_dof_map();
+    std::vector<dof_id_type> dof_indices;
+    dof_map.dof_indices(&elem, dof_indices, variable_num);
+
+    std::vector<double> solution_value(1);
+    system.solution->get(dof_indices, solution_value);
+
+    return static_cast<double>(solution_value[0]);
+}
+
+void SetElementIDasClusterID(libMesh::Mesh &mesh,unsigned int index){
+    for (const auto &elem:mesh.element_ptr_range()){
+        elem->set_extra_integer(index,elem->id());
+    }
+}
+
+bool BelongToCluster(double elem_solution,double neighbor_solution){
+    if (static_cast<int>(elem_solution)==static_cast<int>(neighbor_solution)){
+        //horible idea but for now OK
+        return true;
+    }else{
+        return false;
+    }
+}
+
+void ApplyRecursiveClustering(LinearImplicitSystem& system, const libMesh::Elem& parent_elem, libMesh::Elem& elem, const unsigned int &variable_num, const unsigned int &index) {
+    const libMesh::Elem& parent = parent_elem;
+    if (elem.get_extra_integer(index) == elem.id()) /*failing here */{
+
+        for (unsigned int side = 0; side < elem.n_sides(); side++) {
+            const libMesh::Elem* neighbor = elem.neighbor_ptr(side);
+
+            if (neighbor) {
+                double neighbor_solution = GetElementDataFromMesh(system, *neighbor, variable_num);
+                double elem_solution = GetElementDataFromMesh(system,elem, variable_num);
+
+                if (BelongToCluster(elem_solution,neighbor_solution)) {
+                    //const_cast<libMesh::Elem*>(neighbor)->set_extra_integer(index, elem.get_extra_integer(index));
+                    const_cast<libMesh::Elem*>(neighbor)->set_extra_integer(index, elem.get_extra_integer(index)); //another horible idea but I am gonna blame paraview here
+                    ApplyRecursiveClustering(system, parent, const_cast<libMesh::Elem&>(*neighbor), variable_num, index);
+                }
+            }
+        }
+    }
+}
+
+const unsigned int FindCluster(libMesh::Mesh &mesh, LinearImplicitSystem &system, const std::string &variable_name) {
+    const unsigned int variable_num = system.variable_number(variable_name);
+    const unsigned int index = mesh.add_elem_integer(variable_name);
+
+    SetElementIDasClusterID(mesh, index);
+
+    for (const auto &elem : mesh.element_ptr_range()) {
+        if (elem->get_extra_integer(index) == elem->id()) { // If yes, then they aren't part of any cluster
+
+            for (unsigned int side = 0; side < elem->n_sides(); side++) {
+                const libMesh::Elem* neighbor = elem->neighbor_ptr(side);
+                if (neighbor) {
+                    ApplyRecursiveClustering(system, *elem, const_cast<libMesh::Elem&>(*neighbor), variable_num, index);
```

Comment:
I think you need to run `BelongToCluster()` first.  You don't need to go down the recursive path unless the `elem` and `neighbor` belong to a cluster

Reply:
Or you could maybe get rid of the loop of neighbors here and just call `ApplyRecursiveClustering` with `parent = NULL` and `neighbor = elem`....